File: rightSideViewBinaryTree.py
from queue import Queue
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    """
    TC: O(total number of nodes in the tree) = O(n)
    SC: O(max. breadth of the tree)
    """

    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:

        if root == None:
            return None

        output = []
        queue = Queue()
        queue.put(root)

        while not queue.empty():
            size = queue.qsize()

            for i in range(size):
                front = queue.get()

                if i == size - 1:
                    output.append(front.val)

                if front.left != None:
                    queue.put(front.left)

                if front.right != None:
                    queue.put(front.right)

        return output

File: RightSideViewBinaryTree.java
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    // intialize the return list
    List<Integer> result = new ArrayList<>();
    public List<Integer> rightSideView(TreeNode root) {
        // call the recursive function
        dfs(root, 0);
        return result;
    }
    private void dfs(TreeNode root, int level){
        //base
        // if leaf node in hit then return
        if(root == null) return;
        // logic
        // if level does not exist then add node's value to the list
        if(result.size() == level) result.add(root.val);
        // call recursive function on the right child node first
        dfs(root.right, level + 1);
        // then call recursive function on the left child node
        dfs(root.left, level + 1);
    }
}